application/common/models/User.php

```diff
@@ -0,0 +1,190 @@
+<?php
+
+namespace common\models;
+
+use common\helpers\Utils;
+use Exception;
+use yii\behaviors\AttributeBehavior;
+use yii\db\ActiveRecord;
+use yii\helpers\ArrayHelper;
+
+class User extends UserBase
+{
+    const SCENARIO_SAVE            = 'save';
+    const SCENARIO_UPDATE_PASSWORD = 'update_password';
+    const SCENARIO_AUTHENTICATE    = 'authenticate';
+
+    public $password;
+
+    public function scenarios(): array
+    {
+        $scenarios = parent::scenarios();
+
+        $scenarios[self::SCENARIO_DEFAULT] = [
+            'employee_id',
+            'first_name',
+            'last_name',
+            'display_name',
+            'username',
+            'email',
+            'active',
+            'locked',
+        ];
+
+        $scenarios[self::SCENARIO_UPDATE_PASSWORD] = ['password'];
+
+        $scenarios[self::SCENARIO_AUTHENTICATE] = ['username', 'password'];
+
+        return $scenarios;
+    }
+
+    public function rules(): array
+    {
+        return ArrayHelper::merge([
+            [
+                'active', 'default', 'value' => 'yes',
+            ],
+            [
+                'locked', 'default', 'value' => 'no',
+            ],
+            [
+                ['active', 'locked'], 'in', 'range' => ['yes', 'no'],
+            ],
+            [
+                'email', 'email',
+            ],
+            [
+                'password', 'required',
+                'on' => [self::SCENARIO_UPDATE_PASSWORD, self::SCENARIO_AUTHENTICATE],
+            ],
+            [
+                'password', 'string',
+            ],
+            [
+                'password',
+                $this->validatePassword(),
+                'on' => self::SCENARIO_AUTHENTICATE,
```

Comment:
I'm still having some trouble wrapping my mind around exactly what's happening when we use these validation rules and specific scenarios for things like validating a username and password. These validation rules still feel like they were intended for creating/updating database records, and while we technically can use them for validating a given username and password, it seems to make it harder to clearly see exactly what's going on... and validating a login is one of those situations where I keep wanting to be able to see exactly what's happening, as clearly as possible. Thoughts?